### PR TITLE
Shooting logic supports immutable decker object

### PR DESCRIPTION
**Pull request by overriderb**
Changes after making decker immutable

### CODE REVIEW

File: src/main/scala/battleship/GameEngine.scala [hunk 1]
```diff
@@ -3,19 +3,24 @@ package battleship
...
-      Round(attackingPlayer, defendingPlayer, hitTarget, successful = false)
-    }
+    val hitDefendingPlayer = maybeShip.map(ship =>
+      defendingPlayer.copy(map =
+        defendingPlayer.map.copy(ships =
+          defendingPlayer.map.ships.map {
+            case `ship` => ship.copy(deckers = ship.deckers.map {
+              d => if (d.position == hitTarget) d.copy(alive = false) else d
+            });
+            case x => x
```
Comment (vitold) [inline src/main/scala/battleship/GameEngine.scala]: looks like a redundant expression.
it also makes sense to split this nested logic to make it more clear and testable.

```
    val hitDefendingPlayer = maybeShip.map(ship =>
      defendingPlayer.copy(map =
        defendingPlayer.map.copy(ships =
          defendingPlayer.map.ships
            .map(setDeckers(_, hitTarget)))))

    def setDeckers(ship: Ship, hitTarget: Coordinate): Ship = {
      ship.copy(deckers = ship.deckers.map {
        d => if (d.position == hitTarget) d.copy(alive = false) else d
      })
    }
```

Comment (vitold) [reply src/main/scala/battleship/GameEngine.scala]: To make such transformations much easier: http://julien-truffaut.github.io/Monocle/
This is an implementation of the haskell's lenses.

Comment (overriderb) [reply src/main/scala/battleship/GameEngine.scala]: Looks simpler, thanks
here eaca855

Comment (overriderb) [reply src/main/scala/battleship/GameEngine.scala]: @vitold 
Used monocle for this case, please take a look: 0d8d2c9

Comment (vitold) [reply src/main/scala/battleship/GameEngine.scala]: looks good, but could be even better:

to reduce boilerplate you could use @Lenses("_") annotations to generate lenses
without manual lenses and traversal generation you just need next lines:

```
(Player._map composeLens
      BattleMap._ships composeTraversal each composeLens
      Ship._deckers composeTraversal each) modify(tryHitDecker(_, hitTarget))
```
and don't forget to add compiler pluging for macro paradise:
```
addCompilerPlugin("org.scalamacros" %% "paradise" % "2.1.0" cross CrossVersion.full)
```

Comment (overriderb) [reply src/main/scala/battleship/GameEngine.scala]: Done here 6c9ca6e

**Pull request merged**